scratchpad/issue-258.md

```diff
@@ -1,34 +1,78 @@
+    - `fn read_dir(&self, p: &Path) -> io::Result<Vec<(PathBuf, bool)>>` where the bool indicates is_dir.
+  - Implement `RealFs` using `std::fs::read_dir` and `file_type().is_dir()`; map common errno values to existing `Error` variants (NoPermission, Io, etc.).
+  - Update mock/test FS to support directory listing consistent with the new method. Ensure it can represent both files and directories.
+  - Implement `Resctrl::cleanup_all(&self)`:
+    - If `auto_mount` is true, call `ensure_mounted()`.
+    - If root does not exist and `auto_mount` is false, return `Error::NotMounted`.
+    - List immediate children of the configured root; filter to entries that are directories AND whose name starts with `self.cfg.group_prefix`.
+    - For each candidate, call `FsProvider::remove_dir` and increment a counter on success. On `ENOENT`, skip (idempotent). On other errors, let the library continue (best-effort) and bubble a summary only if the root listing failed at the start.
+    - Return `Ok(removed_count)`.
+  - Keep sanitization minimal: only strict prefix match; do not attempt to parse UIDs here. Rely on group creation to sanitize UIDs.
 
-## Implementation Notes
-- Use strict prefix match and simple sanitization to avoid accidental deletion.
-- Consider a dry-run option or debug logging of candidates.
+- nri-resctrl-plugin
+  - Add an internal `did_startup_cleanup: AtomicBool` (or `OnceCell`) to `ResctrlPlugin`.
+  - On first `synchronize()` when `cfg.cleanup_on_start` is true, call `resctrl.cleanup_all()` and set the flag. If it returns `Error::NotMounted`, log at debug and proceed; otherwise log removed count at info/debug. Never emit `PodResctrlEvent` for cleanup-only actions.
+  - Unit-test this behavior using the plugin’s test FS and DI constructor (`with_resctrl` / `with_pid_source`).
 
-## Dependencies
-- Sub-Issue 01 (crate), Sub-Issue 03 (plugin skeleton).
+## Detailed Test Plan
 
-## Testing
-- Mock FS directory listing of resctrl root; ensure only expected entries removed.
+- resctrl crate unit tests (using mock `FsProvider`):
+  - Root not mounted, `auto_mount=false`: `cleanup_all()` returns `Error::NotMounted` and removes nothing.
+  - Root not mounted, `auto_mount=true`: mock mount succeeds; `cleanup_all()` proceeds and removes matching dirs.
+  - Mixed entries: under `/sys/fs/resctrl`, create entries `pod_a`, `pod_b`, `info` (dir), `tasks` (file), `other` (dir). Only `pod_*` are removed; others remain. Assert returned count is 2.
+  - Permission denied on one directory: `remove_dir` for `pod_b` returns `EACCES`; method still returns `Ok(1)` (best-effort) and other entries remain untouched.
+  - Idempotency: calling `cleanup_all()` twice removes the same initial set once; second call returns 0 and is not an error.
 
-## Risks
-- Accidental deletion mitigated by strict prefix filter and tests.
+- nri-resctrl-plugin tests:
```

Comment:
please add a description of an end-to-end test that we'll add to integration_test.rs, which creates resctrl groups both at the root and under `mon_groups`, these resctrl groups are both `pod_*` and non-prefix, and we test that after starting the plugin, it cleans up the right directories and doesn't touch others.